concourse/steps/release.py

```diff
@@ -23,22 +30,392 @@
+    an object (e.g.: a `dict`) that is then made available to later steps
+    when part of a `Transaction`.
+
+    Subclasses *must* overwrite the `revert` method, which reverts any persistent external side
+    effects previously created by running the step's `apply` method. This should take into account
+    that the execution of the `apply` method may or may not have succeeded, failed,
+    or failed partially.
+    '''
+    def set_context(self, context: TransactionContext):
+        self._context = context
+
+    def context(self):
+        return self._context
+
+    def validate(self):
+        pass
+
+    @abc.abstractmethod
+    def apply(self):
+        return None
+
+    @abc.abstractmethod
+    def revert(self):
+        pass
+
+    @abc.abstractmethod
+    def name(self):
+        pass
+
+
+class Transaction(object):
+    '''This class represents a transaction using `TransactionalStep`s
+
+    Provides aptly named methods to `validate` a series of `TransactionalStep`s and `execute`
+    it atomically, performing the necessary undo actions should an error occur.
+
+    `TransactionalStep`s are provided with access to a shared `TransactionContext` instance
+    to store and retrieve (by step name) information with greater scope than a single step.
+    '''
+    def __init__(
+        self,
+        *steps: TransactionalStep,
+    ):
+        # create context object for this transaction
+        self.context = TransactionContext()
+        # validate type of args and set context
+        for step in steps:
+            if not isinstance(step, TransactionalStep):
+                raise RuntimeError('Transactions may only contain instances of TransactionalStep')
+            step.set_context(self.context)
+        self._steps = steps
+
+    def validate(self):
+        for step in self._steps:
+            info(f"Validating step '{step.name()}'")
+            step.validate()
+
+    def execute(self):
+        executed_steps = list()
+        for step in self._steps:
+            step_name = step.name()
+            # attempt to execute the step
+            try:
+                info(f"Executing step '{step_name}'")
+                executed_steps.append(step)
+                output = step.apply()
+                self.context.set_step_output(step_name, output)
+            except BaseException as e:
+                warning(f"An error occured while executing step '{step_name}': {e}")
+                # revert the changes attempted, in reverse order
+                self._revert(reversed(executed_steps))
+
+    def _revert(self, steps):
+        # attempt to revert each step. Raise an exception if not all reverts succeeded.
+        all_reverted = True
+        for step in steps:
+            step_name = step.name()
+            try:
+                info(f"Reverting step {step_name}")
+                s.revert()
+            except BaseException as e:
+                all_reverted = False
```

Comment:
this is in general fine. it might help us to also dump the stacktrace in these cases (`traceback.print_exc()`). Admittedly, this is not very sophisticated. but it tends to be helpful for error analysis..